Flexible gplates velocities

ebredow opened the pull request:
I finally managed to clean up the GPlates plugin in the style of the AsciiData plugin (with Rene's help). The most important points are: 

The plugin can now also be used for regional models where the gplates velocities are prescribed at the side boundaries up to the lithosphere thickness. 

The new structure solves a problem mentioned by @SiqiZhang last year, where the plate got deformed during the interpolation. 

Unfortunately, all parameter files will need some changes due to the renaming of some input parameters, but the structure is now the same as for the AsciiData plugin and therefore more consistent and also more flexible. 

Also the tests have changed, because the interpolation of the velocities is now being performed in spherical coordinates instead of cartesian coordinates.

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -44,6 +46,7 @@ namespace aspect
        * values at one particular time are interpolated between the two
```
This comment is not longer true. We only keep one file around in GPlatesLookup, but GPlates has two GPlatesLookup objects and interpolate between them (no need to mention that here).

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -54,8 +57,7 @@ namespace aspect
            * parameters for a 2D model.
```
Remove the part about the pointers.

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -63,55 +65,35 @@ namespace aspect
            * SimulatorAccess is not initialized and only Rank 0 should give
```
I think this is not longer true as well. We now have SimulatorAccess during the call to GPlates::initialize(). Remove the second sentence.

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -126,16 +108,6 @@ namespace aspect
           Tensor<2,3> rotation_matrix;
```
Could you clarify the comment that it is not used for 3D *models*.

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -194,35 +166,12 @@ namespace aspect
            * (theta, phi)
            * @param s_position Position in spherical coordinates
-           * (theta,phi,radius)
+           * // alt:(theta,phi,radius)
```
Remove this line

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -254,21 +203,6 @@ namespace aspect
           arc_distance(const Tensor<1,3> position_1, const Tensor<1,3> position_2) const;
```
This function does not any longer get called. I think you can safely remove it.

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -333,55 +267,66 @@ namespace aspect
         /**
          * Time in model units (depends on other model inputs) between two
-         * velocity files.
+         * data files.
```
I think this was my fault copying the stuff from ascii data. Could you replace this with velocity again?

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -333,55 +267,66 @@ namespace aspect
 
         /**
-         * Weight between velocity file n and n+1 while the current time is
+         * Weight between data file n and n+1 while the current time is
```
same here

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -333,55 +267,66 @@ namespace aspect
          * State whether we have time_dependent boundary conditions. Switched
-         * off after finding no more velocity files to suppress attempts to
-         * read in new files.
+         * off after finding no more data files to suppress attempts to read
```
and here

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -333,55 +267,66 @@ namespace aspect
         bool time_dependent;
 
         /**
+         * Directory in which the gplates velocity are present.
```
... gplates velocity files ...

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -403,33 +348,36 @@ namespace aspect
-         * window' interpolation without distance weighting).
+         * Determines the depth of the lithosphere. The user might want to apply
+         * the GPlates velocities not only at the surface of the model, but also
+         * in the whole lithosphere.
```
Explain a bit more what is happening there. Something like: At every side boundary point with a depth smaller than this value the surface velocity will be prescribed.

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -403,33 +348,36 @@ namespace aspect
 
         /**
          * Handles the update of the velocity data in lookup.
          */
```
Can you document the input parameter. It is there to be able to reload both data files (tn and tn+1) if the model time step is larger than the data file time step.

gassmoeller commented on source/velocity_boundary_conditions/gplates.cc:
```diff
@@ -191,21 +177,13 @@ namespace aspect
-        (*velocity_values).reinit(n_theta,n_phi);
-        velocity_positions.reinit(n_theta,n_phi);
+        /**
+         * Tables which contain the velocities
```
Can you clarify in this comment, that is holds the velocities at every point, and that we have two tables, one for the theta component and one for the phi component? By the way, which one is which? I do not remember, we should have documented that right away ;-)

gassmoeller commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -63,55 +65,35 @@ namespace aspect
            */
-          Table<2,Tensor<1,3> > *velocity_values;
-          Table<2,Tensor<1,3> > *old_velocity_values;
+          std_cxx11::array<std::pair<double,double>,2> grid_extent;
```
We only use this variable in GPlatesLookup::load_file(). There is no need to keep it as a member variable in the header file. Just move it to the load_file() function where it is used.

gassmoeller commented on source/velocity_boundary_conditions/gplates.cc:
```diff
@@ -220,207 +198,102 @@ namespace aspect
-        else
-          output_boundary_velocity = convert_tensor<3,dim>(interpolated_velocity);
+        //remove the radius (first entry of internal_position_in_spher_array)
+        Point<2> internal_position_in_spher_rad;
```
Coud you extend the comment by stating that we also sort the components of the spherical position differently? I think it is from [r,phi,theta] to [theta,phi], but better check.

gassmoeller commented on source/velocity_boundary_conditions/gplates.cc:
```diff
@@ -220,207 +198,102 @@ namespace aspect
-
-        const Tensor<1,3> normalized_position = position / position.norm();
+        //transform interpolated_velocity in cartesian coordinates
+        Tensor<1,3> interpolated_velocity_in_cart;
```
you can combine the following two lines and make interpolated_velocity_in_cart `const`

gassmoeller commented on source/velocity_boundary_conditions/gplates.cc:
```diff
@@ -454,8 +328,9 @@ namespace aspect
 
+      template <int dim>
       double
-      GPlatesLookup::arc_distance(const Tensor<1,3> position_1, const Tensor<1,3> position_2) const
```
This function can be removed

gassmoeller commented on source/velocity_boundary_conditions/gplates.cc:
```diff
@@ -741,19 +590,60 @@ namespace aspect
+        }
+      else
+        AssertThrow (false,ExcMessage ("This gplates plugin can only be used when using "
+                                       "a spherical shell or box geometry."));
```
We do not allow box geometries any longer. But instead you should add a ` || (dynamic_cast ... Chunk<dim> ... != 0)` to the if statement above. The spherical chunk should work with no problems with the new gplates plugin.

gassmoeller commented on source/velocity_boundary_conditions/gplates.cc:
```diff
@@ -810,81 +711,58 @@ namespace aspect
       // Give warning if first processor
       this->get_pcout() << std::endl
-                        << "   Loading new velocity file did not succeed." << std::endl
+                        << "   Loading new data file did not succeed." << std::endl
```
Same problem again, copy-paste error of myself. Please change back to velocity.

gassmoeller commented on source/velocity_boundary_conditions/gplates.cc:
```diff
@@ -990,25 +879,26 @@ namespace aspect
-              time_step                *= year_in_seconds;
-              velocity_file_start_time *= year_in_seconds;
+              data_file_time_step        *= year_in_seconds;
+              first_data_file_model_time *= year_in_seconds;;
```
double ;;

gassmoeller commented on tests/gplates-1-3/.directory:
```diff
@@ -0,0 +1,4 @@
+[Dolphin]
```
Please remove the .directory files. They got added by your file explorer.

gassmoeller commented on tests/gplates-1-4/.directory:
```diff
@@ -0,0 +1,4 @@
+[Dolphin]
```
This one as well

gassmoeller commented on tests/gplates_1_3/screen-output:
```diff
@@ -1,13 +1,6 @@
 
 Number of active cells: 768 (on 4 levels)
-Number of degrees of freedom: 10,656 (6,528+864+3,264)
+Number of degrees of freedom: 10.656 (6.528+864+3.264)
```
This is a trick one. Your shell outputs numbers in the german numbering convention. I can simply overwrite the output with results from my machine after all other changes are done.

gassmoeller commented on tests/gplates_1_3/screen-output:
```diff
@@ -38,7 +31,7 @@ Number of degrees of freedom: 10,656 (6,528+864+3,264)
 
    Postprocessing:
-     RMS, max velocity:                  0.0256 m/year, 0.102 m/year
+     RMS, max velocity:                  0.0257 m/year, 0.0769 m/year
```
This is quite a change in max velocity. Before merging we need to check the new version is indeed more accurate than the old one.

bangerth replied:
Indeed. Did you figure out the reason?

ebredow replied:
The reason for the change in max velocity is probably that we now interpolate differently than before (in spherical coordinates instead of cartesian coordinates). If it is correct that this change is so big is rather difficult to verify. I discussed this issue with Rene and checked two points: 
1. In the visualization of the test outputs, the velocities at the surface point only in phi or theta direction. 
2. If I load the current day velocities as input in the test cases, the velocities at the surface seem to be in the usual range of present-day plate velocities. 
So far, we couldn't think of another way to find out more about that.

bangerth replied:
You're on a very coarse mesh, so the difference in interpolation may well make the difference. It would be worth seeing whether you get smaller differences from this testcase if you run it with a once-refined mesh with and without your patches.

ebredow replied:
I have run the tests now for a once-refined and also a twice-refined mesh, but unfortunately, the differences before and after the patches remain almost the same. The numbers for the max. velocity in both versions get both slightly higher, so the difference stays more or less the same.

gassmoeller commented:
Hi Eva, thanks for opening the PR. I made a number of comments, mostly concerning code style and improving documentation, but apart from that I am fine with the result (as you know ;-)). Please address them (or ask for help, when you get stuck), and let me know when you are finished. 
After that, I would like to ask somebody else to go over this PR, because it is changing some functionality and I was working a lot on it, therefore I am not an independent reviewer.

gassmoeller commented:
/run-tests

ebredow commented:
@gassmoeller:  Hi Rene, as usual, this was not as simple as expected. I addressed all of your comments and the change in the max .velocity seems to be fine (I checked it as we discussed last week). The only thing left to do are to switch points and commas in the test files. 

Additionally, I cleaned up a little bit more: I also removed the functions "rotate_grid_velocity", "get_grid_point_position" and "rotate_around_axis" as they are no longer called anywhere (and the tests don't change). Do you agree to that?

And then things got a little bit weird with the tests. The numbers have not changed, but the order of the statements in the screen-output files. I think this must have happened when I rebased my branch to the master branch. I probably did not run the tests after the rebase and before opening the PR - because if I had, I would have seen this problem earlier. Since the numbers have not changed, I overwrote the screen-output files. 
One question is left however: we did no longer call the screen-output function in the last version and therefore, the screen output disappeared. I now put the call in initialize() and this works fine, but it is probably not the best place - do you remember why we removed it in update() and should we move it there again?

Cheers from Potsdam =)

gassmoeller commented:
Ok, this should be ready for a review by somebody else than me, I had the idea and guided Eva during the coding, so I am biased.

bangerth commented:
@ebredow , two questions about backward compatibility and changes in behavior:

* Can you explain in the changes.h file how exactly one had to adjust input files? 
* What is the impact of doing interpolation in spherical coordinates (as opposed to doing so in Cartesian coordinates) when you are close to the poles? Does interpolation continue to work correctly and accurately there?

bangerth commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -38,80 +40,47 @@ namespace aspect
-       * instances in time where GPlates provides us with data; the boundary
-       * values at one particular time are interpolated between the two
-       * currently loaded data sets.
+       * velocity boundary condition from a gplates .gpml file.
```
Is the previous description no longer the case? How do you do interpolation in time now?

ebredow replied:
The previous description is no longer correct. The GPlatesLookup class keeps only one file around, but the GPlates class has two GPlatesLookup objects and interpolates between them. Rene thought it would not be necessary to mention that here, but maybe I should add it to avoid further confusion.

ebredow commented:
* Of course I can add an entry in the changes.h file to clarify the changes in the input files. I will do so tomorrow when I have my laptop at the office.
* As for the interpolation in spherical coordinates close to the poles - this is a very good point that we did not test so far. I will do that also tomorrow. 

@bangerth: Thanks for looking at this =)

ebredow commented:
I am not sure if rebasing to the master branch was the right thing to do, but as I modified changes.h, I wanted to apply the changes to the current version of this file. Anyway, the actual new commits are the three short ones at the end.

ebredow commented:
OK, now the interpolation in spherical coordinates close to the poles works fine for all our tests (current_day, theta_test, phi_test). 
Additionally we fixed a bug to make sure that the velocities are applied at all points on the surface independent of the choice for the lithosphere thickness.

ebredow commented:
OK, the rebase conflicts are resolved and the tester seems to be happy, so I guess this is finally ready to be reviewed (again) by someone else than @gassmoeller.

gassmoeller commented:
Ping @bangerth or @tjhei: Could one of you take a look at this PR in the next weeks? We are pretty sure it is more accurate than the old version + it is more similar to all the ascii data plugins. In particular, I would like to merge this before the next release, because the current state is introducing artificial deformations at the surface, when interpolation is enabled (my mistake from May 2014).

bangerth commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -38,80 +40,47 @@ namespace aspect
            */
-          Table<2,Tensor<1,3> > *velocity_values;
-          Table<2,Tensor<1,3> > *old_velocity_values;
+          std::vector<Functions::InterpolatedUniformGridData<2> *> velocities;
```
Can you make this a vector of `std::shared_ptr` or `std::unique_ptr`? This way you won't have to deal with manually managing memory.

bangerth replied:
As a matter of fact, you know that you're only going to have 2 elements, so you may as well use a `std_cxx11::array<2,std_cxx11::unique_ptr<...>>` here.

bangerth commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -333,40 +222,39 @@ namespace aspect
+         * gets updated if necessary by update().
          */
-        double time_relative_to_vel_file_start_time;
+        int  current_file_number;
```
Could we make this an unsigned int (because conceptually a file number is a non-negative number)? If you need an invalid value, you can always use `numbers::invalid_unsigned_int`.

bangerth commented on include/aspect/velocity_boundary_conditions/gplates.h:
```diff
@@ -333,40 +222,39 @@ namespace aspect
+         * larger than 'First data file model time'.
          */
-        double velocity_file_start_time;
+        int first_data_file_number;
```
same here

bangerth commented on source/utilities.cc:
```diff
@@ -376,7 +385,7 @@ namespace aspect
       std::ifstream in(filename.c_str(), std::ios::in);
       AssertThrow (in,
-                   ExcMessage (std::string("Could not open data file <"
+                   ExcMessage (std::string("Couldn't open data file <"
```
I think I've converted a number of places to more formally use "could not" instead of "couldn't".

bangerth commented on source/velocity_boundary_conditions/gplates.cc:
```diff
@@ -191,26 +176,21 @@ namespace aspect
+         * velocity_values[0] is the table for the theta component, whereas
+         * velocity_values[1] is the table for the phi component.
+         */
+        std::vector<Table<2,double> > velocity_values(2,Table<2,double>(n_theta,n_phi));
```
You could convert this into a fixed size array using `std_cxx11::array` if you feel ambitious, but it doesn't matter much.

bangerth replied:
Well, or just
```
  Table<2,double> velocity_values[2] = {..., ...};
```

gassmoeller replied:
Is the initialization list a C++11 feature? If so we should use the `std::cxx11::array`.

bangerth replied:
Initializer lists are C++11, but initialization of arrays via `{...,...}` is old-style K&R C from 1977 or so ;-)

bangerth commented on source/velocity_boundary_conditions/gplates.cc:
```diff
@@ -191,26 +176,21 @@ namespace aspect
         std::stringstream in(velos, std::ios::in);
         AssertThrow (in,
-                     ExcMessage (std::string("Could not find velocities. Is this file in native gpml format for the velocities?")));
+                     ExcMessage (std::string("Couldn't find velocities. Is file native gpml format for velocities?")));
```
Same as above, the grammar of the new error message sounds less formal than before. The second sentence also is grammatically wrong.

bangerth commented:
This looks reasonable to me.

The only real question is whether we're sure that the differences in output are a result of being more correct, rather than having new bug introduced. What's your take?

Other than that, just address my few little niggles, rebase, and merge.

bangerth commented:
This does not currently compile because we only require deal.II 8.2 which did not have `unique_ptr`:
```
/aspect-tester/aspect/include/aspect/velocity_boundary_conditions/gplates.h:27:47: fatal error: deal.II/base/std_cxx11/unique_ptr.h: No such file or directory
 #include <deal.II/base/std_cxx11/unique_ptr.h>
                                               ^
```

Can you just use `shared_ptr` instead?

ebredow commented:
I think this is finally being ready to be merged - sorry it took so long!

gassmoeller commented:
Since all tests are working and the comments have been addressed I will merge this now. In case we find new issues we should handle them in a new pull request, this one has been open for far too long. Thanks for fixing this @ebredow!

bangerth commented:
Indeed, thanks for sticking with us through all of these revisions!
